Data/PackageData.cs

```diff
@@ -404,12 +403,11 @@ private static async Task<PackageData> ReadPackageFromUrl (PackageData package,
             {
                 //System.Console.WriteLine($"DOWNLOADING {package.DownloadUrl}");
                 var r = await httpClient.GetAsync (package.DownloadUrl, token).ConfigureAwait (false);
-                var data = new MemoryStream ();
-                using (var s = await r.Content.ReadAsStreamAsync ().ConfigureAwait (false)) {
-                    await s.CopyToAsync (data, 16 * 1024, token).ConfigureAwait (false);
+                using (var content = r.Content) {
+                    using (var s = await content.ReadAsStreamAsync ().ConfigureAwait (false)) {
```

Comment:
Keep in mind that packages on nuget.org can be up to 250MB. Your change avoids a copy but may still buffer the entire stream into memory (also, this may break later if `Read` requires a seekable stream).

On nuget.org a common strategy we use is to download the package to disk, and then use a `FileStream` for read operations. I would recommend a similar approach here.

Reply:
Thanks for the feedback. This was mainly a quick-fix to try and reduce the occurrences of the errors, as they don't seem to happen often from what I've seen. The problem isn't so much the amount of memory required, but the way that it's allocated. A memory stream (similar to a List) will double itself every time it gets full, and once it gets to 85k it'll be on the large object heap which (by default) doesn't compact and defrag. This means that to take your example of 250mb there will be 12 "gaps" left on the large object heap and the process will have allocated nearly 512mb of RAM. Removing this copy alleviates all of that pressure.

I do take your point about seekable streams though. This is dependent on `ReadAsStreamAsync` not changing behaviour (as it is already providing a seekable stream)

Reply:
> Removing this copy alleviates all of that pressure.

Removing the copy will reduce the memory pressure, but does not remove it entirely as the `HttpClient.GetAsync` method buffers the content in a `MemoryStream` (see [this](https://github.com/dotnet/runtime/blob/85ba3e19240c453362381cd89db618830f7951ad/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L517-L534)) when you use the default `HttpCompletionOption.ResponseContentRead`. However, you can avoid the problem by spilling to disk using the `FileStream` approach I suggested earlier.

Reply:
Thanks for the suggestion. I've had a look at the code that you've pointed me to but I don't think writing to disk gains much, if anything. If I'm reading it right, the method you use still calls `ReadAsStreamAsync` which means it's still getting buffered into a `MemoryStream` before being written to disk, so it will still have the one-off memory pressure. I haven't benchmarked it yet (may try this weekend) but as these are not lengthy operations, it is entirely possible that writing the output to disk could cause the memory to be in use for longer than just doing the processing in memory, depending on IO speed.

Reply:
> ... the method you use still calls `ReadAsStreamAsync` which means it's still getting buffered into a `MemoryStream` ...

The content is not buffered due to the `HttpCompletionOption.ResponseHeadersRead` [here](https://github.com/NuGet/NuGet.Jobs/blob/262da48ed05d0366613bbf1c54f47879aad96dcd/src/Validation.Common.Job/FileDownloader.cs#L73). This results in minimal memory pressure even if you are downloading a 250mb package.

Reply:
Ah I see now. I'd missed that that option changed the behaviour of the stream created. I'll modify this PR to copy to disk first as that should completely resolve the bugs mentioned above